feat: Lineage Graph visualization beta

bbsbb opened the pull request:
### Summary of Changes

Initial visualization for table lineage via D3.
No additional configuration is required for this feature. Having table lineage enabled will add a Lineage(beta) button to the main bar on the Table details page which leads to the d3 visualization panel. Releasing this would allow us to get early community feedback on usage and facilitate next iteration.

Visualization: 

https://user-images.githubusercontent.com/6582764/123804896-4c625800-d8f6-11eb-87d2-2abe2f572fc7.mp4




### Tests

Goal is to build tests for the feature when exiting beta stage.

### Documentation

N/A

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [] PR includes a summary of changes.
- [] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

bbsbb commented:
@verdan @Golodhros @dorianj  May I get :eyes: on this, please.

mgorsk1 commented:
Nice! I don't have that much exp with frontend to take this on but I've deployed it locally with some test data and noticed issue when one node (same key) has multiple parents. Following lineage spec (correct as far as I understand our approach) yields in failure in lineage view:

```
{
  "upstream_entities": [
    {
      "usage": 0,
      "parent": "",
      "level": 1,
      "key": "hive_table://demo.ecuador/ray_cole",
      "badges": [],
      "source": "hive_table"
    }
  ],
  "depth": 5,
  "direction": "both",
  "downstream_entities": [
    {
      "usage": 0,
      "parent": "hive_table://demo.poland/floyd_little_and_hunter",
      "level": 2,
      "key": "hive_table://demo.poland/ryan_inc",
      "badges": [],
      "source": "hive_table"
    },
    {
      "usage": 0,
      "parent": "hive_table://demo.poland/floyd_little_and_hunter",
      "level": 2,
      "key": "hive_table://demo.mauritania/king_lee",
      "badges": [],
      "source": "hive_table"
    },
    {
      "usage": 0,
      "parent": "hive_table://demo.poland/ryan_inc",
      "level": 3,
      "key": "hive_table://demo.poland/camacho_bass_and_hernandez",
      "badges": [],
      "source": "hive_table"
    },
    {
      "usage": 0,
      "parent": "hive_table://demo.poland/dyer_inc",
      "level": 1,
      "key": "hive_table://demo.poland/floyd_little_and_hunter",
      "badges": [],
      "source": "hive_table"
    },
    {
      "usage": 0,
      "parent": "hive_table://demo.mauritania/king_lee",
      "level": 3,
      "key": "hive_table://demo.poland/camacho_bass_and_hernandez",
      "badges": [],
      "source": "hive_table"
    }
  ],
  "key": "hive_table://demo.poland/dyer_inc"
}
```

![image](https://user-images.githubusercontent.com/5680655/121240555-c1fc7a80-c89a-11eb-9531-bb535bed623e.png)

Hope this is somewhat useful.

Golodhros commented on frontend/amundsen_application/static/css/_animations.scss:
```diff
@@ -5,6 +5,33 @@
+
+.is-pulse-animated {
+  @extend %is-pulse-animated;
+}
```
Nice!

Golodhros commented on frontend/amundsen_application/static/js/components/LineageGraph/index.tsx:
```diff
@@ -0,0 +1,84 @@
+
+interface LineageGraphState {}
+
+export class LineageGraph extends React.Component<
```
We want to move out of using the class syntax for new work, could you try to use React Hooks instead?

Golodhros replied:
This might be useful to you: https://github.com/britecharts/britecharts-react/blob/mi-typescript-move/src/charts/ComponentDemo.tsx

Golodhros commented on frontend/amundsen_application/static/js/components/LineageGraph/index.tsx:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    // Setting up margins & render screen for the graph.
+    const margins = { top: 20, right: 100, bottom: 20, left: 100 };
```
nit: this could be a constant outside the component

Golodhros commented on frontend/amundsen_application/static/js/components/LineageGraph/styles.scss:
```diff
@@ -0,0 +1,82 @@
+      color: $gray30;
+      width: $svg-label-width;
+      height: $svg-label-height;
+      text-transform: uppercase;
```
Is there a way of using our new standard font styles (https://github.com/amundsen-io/amundsen/blob/main/frontend/amundsen_application/static/css/_typography-default.scss#L17)

Also, I think we will move out of using uppercase for a11y reasons. Probably the designs are not reflecting this, but it was something new that we noticed later.

Golodhros commented on frontend/amundsen_application/static/js/components/LineageGraph/styles.scss:
```diff
@@ -0,0 +1,82 @@
+
+  &.plus {
+    fill: $purple60;
+    font-size: 25px;
```
Let's use a variable for this.

Golodhros commented on frontend/amundsen_application/static/js/pages/LineagePage/index.tsx:
```diff
@@ -0,0 +1,217 @@
+  </div>
+);
+
+export class LineagePage extends React.Component<
```
same comment about class syntax

bbsbb replied:
Refactored all to FC

Golodhros commented on frontend/amundsen_application/static/js/pages/LineagePage/index.tsx:
```diff
@@ -0,0 +1,217 @@
+    // We want to avoid rendering the previous table's metadata before new data is fetched in componentDidMount
+    if (isLoading || !this.didComponentMount) {
+      innerContent = (
+        <div className="pulse-lineage">
+          <div className="pulse-lineage-media is-pulse-animated" />
+        </div>
```
I would extract a tiny component (GraphLoadingState) for this.

Golodhros commented on frontend/amundsen_application/static/js/pages/LineagePage/index.tsx:
```diff
@@ -0,0 +1,217 @@
+              >
+                {rootNodeData.schema}.{rootNodeData.name}
+                <span className="text-secondary lineage-graph-label">
+                  Lineage Graph
```
Let's get all strings into constants at the top of the file, like GRAPH_TITLE or similar

Golodhros commented on frontend/amundsen_application/static/js/pages/LineagePage/index.tsx:
```diff
@@ -0,0 +1,217 @@
+                title={`${rootNodeData.schema}.${rootNodeData.name}`}
+              >
+                {rootNodeData.schema}.{rootNodeData.name}
+                <span className="text-secondary lineage-graph-label">
```
We don't use these classes anymore, now we favor to add an extend to the 'lineag-graph-label' like in https://github.com/amundsen-io/amundsen/blob/main/frontend/amundsen_application/static/css/_typography-default.scss#L150

Golodhros commented on frontend/amundsen_application/static/js/pages/LineagePage/index.tsx:
```diff
@@ -0,0 +1,217 @@
+                  Lineage Graph
+                </span>
+              </h1>
+              <div className="text-body-w2">
```
This is kind or OK, but still better to use a .lineage-graph-back-link class and extend its styles to text-body-w2

Golodhros commented on frontend/amundsen_application/static/js/pages/LineagePage/index.tsx:
```diff
@@ -0,0 +1,217 @@
+      );
+    } else if (statusCode === SERVER_ERROR_CODE) {
+      innerContent = <ErrorMessage />;
+    } else {
```
I would extract everything on this branch into a new component and leave this file as a simple branching component with the connection to the redux store.

This will allow to test it a lot easier.

Golodhros commented on frontend/amundsen_application/static/js/pages/LineagePage/styles.scss:
```diff
@@ -0,0 +1,49 @@
+
+.pulse-lineage-media {
+  height: $pulse-loader-height - $spacer-1;
+  background-image: url('../../../images/icons/Lineage.svg');
```
Would it be possible to make Lineage.svg an extra SVG react component and use it so, without setting it as a background image?

bbsbb replied:
Let's not.
Both animations (shimmer + pulse) are currently background focused. I would rather keep it this way as it is more consistent.

Golodhros replied:
Ack. Sounds good.

Golodhros commented on frontend/amundsen_application/static/js/pages/TableDetailPage/LineageButton/index.tsx:
```diff
@@ -0,0 +1,41 @@
+      href={href}
+      role="button"
+      id="table-lineage-button"
+      onClick={logClick}
```
We are trying to move out of using logClick as the only tracker, using Redux actions and their analytics payload to track.

See an example on https://github.com/amundsen-io/amundsen/blob/main/frontend/amundsen_application/static/js/ducks/bookmark/reducer.ts#L28

You will need to pass down the dispatcher down to this component and probably create an action for it. See (https://github.com/amundsen-io/amundsen/blob/main/frontend/amundsen_application/static/js/ducks/lineage/reducer.ts#L67)

That said, I am wondering why we are following an href here instead of using a router link and avoiding a full reload.

Golodhros commented on frontend/amundsen_application/static/js/components/LineageGraph/wrapperLineage.ts:
```diff
@@ -0,0 +1,290 @@
+      d.y0 = d.y;
+    });
+  }
+};
```
This looks OK, but I don't like the idea of having all the code in one function like this. This makes it really hard to test, extend, and modularize.

I would rather have us creating a graph component based on the Reusable API, which can be tested, configured and extended.

See an example here: https://github.com/britecharts/britecharts-react/blob/mi-typescript-move/src/charts/chartDemo.ts

Golodhros reviewed (commented):
A lot of work, thanks for this!

I added some comments, some of them will need a change of strategy, but I think they are needed for making this testable.

Golodhros commented on frontend/amundsen_application/static/js/components/LineageGraph/index.tsx:
```diff
@@ -0,0 +1,84 @@
+  }
+}
+
+export default LineageGraph;
```
I would love to have tests for this file.

Golodhros commented on frontend/amundsen_application/static/js/pages/TableDetailPage/LineageButton/index.tsx:
```diff
@@ -0,0 +1,41 @@
+  );
+};
+
+export default LineageButton;
```
Could you add a basic test for this component too?

mgorsk1 commented:
When implementing the lineage support in Atlas I've found lineage graph is actually a DAG not a tree, so if d3-hierarchy doesn't support multiple parents for single node then maybe we should use something like https://github.com/erikbrinkman/d3-dag ?

Golodhros commented on frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:
```diff
@@ -0,0 +1,327 @@
+    .attr('class', 'direction-label  upstream-label')
+    .attr(
+      'transform',
+      `translate(${dimensions.width / 2 - 150 - LINEAGE_SCENE_MARGIN.right}, ${
```
That 150 seems like a magic number. Could we create a variable for it?

Golodhros commented on frontend/amundsen_application/static/js/pages/TableDetailPage/LineageLink/index.spec.tsx:
```diff
@@ -0,0 +1,34 @@
+describe('LineageLink', () => {
+  const setup = () => {
+    const props = { tableData: globalState.tableMetadata.tableData };
+    const wrapper = mount(<LineageLink {...props} />);
```
```suggestion
    // eslint-disable-next-line react/jsx-props-no-spreading
    const wrapper = mount(<LineageLink {...props} />);
```

judy-palimonka commented:
Looks great. Very keen on getting this merged soon. Anything I can help with?

bbsbb commented:
Sorry, didn't have time to address feedback over the last week, will be on it addressing the issues in the next days. Thank you all!
@judy-palimonka  - Will make all effort to get it landed within the week!

mgorsk1 commented:
did you manage to resolve the multiple parents issue @bbsbb ?

bbsbb commented:
> did you manage to resolve the multiple parents issue @bbsbb ?

No, but I have a very good gasp on the reasons and I will resolve it and ping you explicitly. Not leaving it like that ;o)

mgorsk1 commented:
Thanks for pinging me @bbsbb I have build the latest frontend lineage version and indeed see some improvement but not sure if the issue was solved 100%.

This is my current lineage:

![image](https://user-images.githubusercontent.com/5680655/123824742-99e2c300-d8fe-11eb-9eb4-c1f9960907b4.png)

1. I start with alvarez_and_sons table view and then to lineage view - works just fine, some minor scaling issues but in general cool stuff!

2. I move to owen_lester table view and then to lineage view - the same 'multiple roots' error appears and the view is not rendered.

I am attaching full flow of what I've described just now, hope it's helpful.

https://user-images.githubusercontent.com/5680655/123825736-73715780-d8ff-11eb-8c38-ca9b327e54e4.mp4

bbsbb commented:
Hey @mgorsk1 ,

Unfortunately, the multiple roots error is not caused by the rendering strategy, but by mismatch between what the LineageItem type in TS expects as mandatory fields, and what is actually available in the response. Could you please enumerate the keys that are **present** in a lineage record from the failing case? (e.g. cluster/database/name etc.).
When you provided the initial data sample above, I assumed that you've manually edited keys out, was that not the case? Is that complete data?

bbsbb commented:
The expectation is that depth 1 entities from both upstream/downstream carry the currently visualized table as parent(which we render as single root). However, I can see how from lineage perspective, this would not be a correct "relationship". Will adjust this expectation to deal with this.

Golodhros commented on frontend/amundsen_application/static/js/components/Lineage/Graph/chart.spec.ts:
```diff
@@ -0,0 +1,140 @@
+  id: '2',
+};
+
+// const nodeWithParent: LineageItem & { parent: LineageItem } = {
+//   ...node,
+//   parent: node,
+// };
```
Code to clean up

Golodhros commented on frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:
```diff
@@ -0,0 +1,428 @@
+    .attr('class', 'plus')
+    .attr('text-anchor', 'middle');
+
+  //
```
To clean?

Golodhros commented on frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:
```diff
@@ -0,0 +1,428 @@
+export const buildSVG = (el: HTMLElement, dimensions: Dimensions) => {
+  const svg = select(el)
+    .append('svg')
+
```
extra space?

Golodhros commented on frontend/amundsen_application/static/js/components/Lineage/Graph/index.tsx:
```diff
@@ -0,0 +1,67 @@
+  // Setting up margins & render screen for the graph.
+  React.useEffect(() => {
+    if (!scene && rootRef.current) {
+      const containerDimensions = rootRef.current.getBoundingClientRect();
+      const dimensions = {
+        width:
+          (containerDimensions.width || CHART_DEFAULT_DIMENSIONS.width) -
+          (LINEAGE_SCENE_MARGIN.left + LINEAGE_SCENE_MARGIN.right),
+        height:
+          (containerDimensions.height || CHART_DEFAULT_DIMENSIONS.height) -
+          (LINEAGE_SCENE_MARGIN.top + LINEAGE_SCENE_MARGIN.bottom),
+      };
```
Nit: I would try to extract this logic into a `getDimensions` function

Golodhros commented on frontend/amundsen_application/static/js/components/Lineage/Graph/index.tsx:
```diff
@@ -0,0 +1,67 @@
+        {
+          upstream: `${ReactDOMServer.renderToString(
+            <div className="text-title-w2">
+              <LeftArrowIcon fill="#acacc0" />
```
Can we put the color in a variable?

Golodhros commented on frontend/amundsen_application/static/js/components/Lineage/Graph/index.tsx:
```diff
@@ -0,0 +1,67 @@
+          upstream: `${ReactDOMServer.renderToString(
+            <div className="text-title-w2">
+              <LeftArrowIcon fill="#acacc0" />
+              <span>Upstream</span>
```
We are trying to get our strings out of the components render function, in this case I would create a constant in this file named UPSTREAM_ARROW_LABEL

Golodhros commented on frontend/amundsen_application/static/js/components/Lineage/Graph/index.tsx:
```diff
@@ -0,0 +1,67 @@
+          )}`,
+          downstream: `${ReactDOMServer.renderToString(
+            <div className="text-title-w2">
+              <RightArrowIcon fill="#acacc0" />
+              <span>Downstream</span>
```
Same comments

Golodhros commented on frontend/amundsen_application/static/js/pages/TableDetailPage/LineageButton/index.tsx:
```diff
@@ -0,0 +1,31 @@
+
+  return (
+    <Link to={path} className="btn btn-default btn-lg">
+      Lineage
```
Let's make this a constant in this file

Golodhros replied:
You forgot to remove the old one?

bbsbb replied:
Indeed, thank you, fixed!

Golodhros commented on frontend/amundsen_application/static/js/pages/TableDetailPage/LineageButton/index.tsx:
```diff
@@ -0,0 +1,31 @@
+    <Link to={path} className="btn btn-default btn-lg">
+      Lineage
+      <span className="static-badge flag label label-warning">
+        <div className="badge-overlay-primary">beta</div>
```
Same here

Golodhros reviewed (commented):
Looking really good!

Just some minor comments, almost ready to merge!

bbsbb commented:
@Golodhros  - All reflected + some additional tests added! 

@mgorsk1  - Handled the no parent case.

Btw, feels a bit awkward that the lineage model is subject to interpretation by different databuilder extractors. Makes me think community will bring up more similar edge cases, but am very much willing to handle on case by case.

Thanks again to both, let me know if anything else is required.

Golodhros reviewed (approved):
LGTM

Thanks a lot for taking this through the finish line!
Great work!!

mgorsk1 commented:
Thanks for fixing this @bbsbb! I'd hardly say this is edge case for lineage (no parent key). We're not really defining something but rather adhering to something that is already defined. I still think that some nodes in the lineage I've showed are non-functional (while they are rather common situations). The next one would be:

```
{
  "downstream_entities": [
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/murphy_taylor_and_rivera",
      "level": 1,
      "name": "murphy_taylor_and_rivera",
      "parent": "hive_table://demo.sri_lanka/howard_harrington_and_mendoza",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    }
  ],
  "upstream_entities": [
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "level": 3,
      "name": "alvarez_and_sons",
      "parent": "",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.togo/morris_plc",
      "level": 1,
      "name": "morris_plc",
      "parent": "hive_table://demo.austria/peterson_plc",
      "schema": "togo",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/owen_lester",
      "level": 2,
      "name": "owen_lester",
      "parent": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/peterson_plc",
      "level": 2,
      "name": "peterson_plc",
      "parent": "",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    }
  ]
}
```

bbsbb commented:
@mgorsk1  - The neo4j lineage extracts I have of lineage all start bidirectional from root with level 1. The flow is:

level 0 root, level 1 immediate parent/child upstream/downstream, level 2 - 1 separated upstream/downstream etc.

In the example you linked, the parent links in the nodes and levels seem direction aware(I made an assumption that the level flow matches and it doesn't seem to be the case) based on if something is upstream or downstream? Could you confirm if the visual you expect would be this:

![2021-06-29-234025_1969x1162_scrot](https://user-images.githubusercontent.com/6582764/123864425-6bcba600-d933-11eb-8985-3c2e657dca03.png)

mgorsk1 commented:
This one essentialy would be:

![image](https://user-images.githubusercontent.com/5680655/123869470-55bce600-d931-11eb-8f27-795757ce3862.png)

Another one buggy is morris-plc:

```
{
  "downstream_entities": [
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/murphy_taylor_and_rivera",
      "level": 2,
      "name": "murphy_taylor_and_rivera",
      "parent": "hive_table://demo.sri_lanka/howard_harrington_and_mendoza",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/howard_harrington_and_mendoza",
      "level": 1,
      "name": "howard_harrington_and_mendoza",
      "parent": "hive_table://demo.togo/morris_plc",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.christmas_island/oliver_davis_and_reid",
      "level": 1,
      "name": "oliver_davis_and_reid",
      "parent": "hive_table://demo.togo/morris_plc",
      "schema": "christmas_island",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/murphy_taylor_and_rivera",
      "level": 2,
      "name": "murphy_taylor_and_rivera",
      "parent": "hive_table://demo.christmas_island/oliver_davis_and_reid",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    }
  ],
  "upstream_entities": [
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "level": 2,
      "name": "alvarez_and_sons",
      "parent": "",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/owen_lester",
      "level": 1,
      "name": "owen_lester",
      "parent": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/peterson_plc",
      "level": 1,
      "name": "peterson_plc",
      "parent": "",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    }
  ]
}

```
that should render

![image](https://user-images.githubusercontent.com/5680655/123869787-be0bc780-d931-11eb-96a3-10d7508c2899.png)



If you have examples on how to render those then feel free to post it I'll try to get what's the expected format of the data, I am pretty sure my responses are correct and I am not really sure what bidirectional means so maybe it would be easier.

bbsbb commented:
@mgorsk1  - Currenlty working to fix upstream rendering and will open a PR in a little bit, but I am slightly perplexed by this picture:
![upstream-lineage](https://user-images.githubusercontent.com/6582764/123919664-d3164400-d98d-11eb-8514-0bd698f784f8.png)


Where does the relationship owen_lester -> morris_plc come from? Isn't there a parent<-> child relationship missing, shouldn't the upstream include an additional record to indicate the 2nd parent, like this:

```javascript
[
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "level": 3,
      "name": "alvarez_and_sons",
      "parent": "",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.togo/morris_plc",
      "level": 1,
      "name": "morris_plc",
      "parent": "hive_table://demo.austria/peterson_plc",
      "schema": "togo",
      "source": "hive_table",
      "usage": 0
    },
     {                                                                                                                                                                                      
       "badges": [],                                                                                                                                                                          
       "cluster": "demo",                                                                                                                                                                     
       "database": "hive_table",                                                                                                                                                              
       "key": "hive_table://demo.togo/morris_plc",                                                                                                                                            
       "level": 1,                                                                                                                                                                            
       "name": "morris_plc",                                                                                                                                                                  
       "parent": "hive_table://demo.austria/owen_lester",                                                                                                                                     
       "schema": "togo",                                                                                                                                                                      
       "source": "hive_table",                                                                                                                                                                
       "usage": 0                                                                                                                                                                             
     }, 
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/owen_lester",
      "level": 2,
      "name": "owen_lester",
      "parent": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/peterson_plc",
      "level": 2,
      "name": "peterson_plc",
      "parent": "",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    }
  ]
```javascript

Thank you again, you have been invaluable in fixing this for good.

mgorsk1 commented:
Thanks for replying. I think that it's implicit that upstream entities of depth 1 are the parents of the given node. 



My understanding is that `upstream` are all tables `before` table X for which the table is retrieved. If we have a table X for which the lineage is retrieved, should this table ever appear in "key" attribute of upstream entities ? 

And with that in mind I don't really see how it makes sense to have this in upstream:
```
{
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.togo/morris_plc",
      "level": 1,
      "name": "morris_plc",
      "parent": "hive_table://demo.austria/peterson_plc",
      "schema": "togo",
      "source": "hive_table",
      "usage": 0
    }
```

our model indeed might be slightly confusing and ambigous in figuring out, that's why I am trying to sort this out from the get go as it impacts implementation on other proxies like Atlas.

bbsbb commented:
I agree with you entirely on the implicit implication in terms of depth 1 - makes a lot of sense as we are focused around a single root - this is how we do it right now as well.

However, I do not believe this would hold for <any depth> + 1(e.g. as soon as there is more than 1 node at a certain level, the "implicit" breaks down as a decision needs to be made to which of the N nodes from previous level to tie to).

Furthermore, such an implementation makes the upstream/downstream parts of our API response inconsistent.
Here is how we treated the same relationship downstream from the posts above:
![downstream_example](https://user-images.githubusercontent.com/6582764/123922350-9e57bc00-d990-11eb-8bd6-df476e8e8d3a.png)

I would appreciate if we can bring the convo to the Amundsen slack, when you have 5 minutes, so we can figure out the desired implementation together.
